### DISCUSSION TITLE
Soulseek protocol extensions

### DISCUSSION THREAD

**Discussion by jpdillingham** [nicotine-plus/nicotine-plus#1772] [General]
I hazily recall something about either a Nicotine+ or Museek+-specific message code, and I (less hazily) recall a discussion about how it isn't desirable to extend the protocol for the purposes of one client.

I'm toying with the idea of doing this, but it feels less offensive if more than one client supports it.

Specifically, I'm testing concurrent transfers to a single user at the moment and am finding that Qt and Nicotine+ handle them just fine while NS does not; it appears to be limited to exactly one transfer connection per user.  I'd love to be able to support concurrent transfers, as there's very clear benefit to doing this (up to a point), but enabling it would outright break functionality for NS users.

That is, unless there were a protocol extension that clients could use to negotiate features.

What I'm thinking is that we could introduce peer message code 5000 with no request body, and then clients that support protocol extensions would respond with a 64 bit integer representing up to 64 flags.  Codes beyond 5000 could then be used to support enabled protocol extensions. In this scenario we'd need a single source of truth regarding what each bit of the extension mask represented, and some sort of registry of codes and message contents.

I just "scanned" Soulseek Qt and confirmed that codes 52-5500 are unused (I didn't bother to go higher), so somewhere in that range would be safe for extension.  I ran the same test against NS and Nicotine+, and neither client had any problem with it.

The only technical challenge I see with this is that unsupported clients won't respond to a code 5000, and clients that are probing for it won't know whether the network is being slow or the client doesn't support extensions.  This can be handled with a timeout, but all new connections (caching notwithstanding) would then need ~5 seconds to negotiate extensions prior to doing work.

I was hoping I could find a peer ping (part of the reason I did the code "scan")  so that a message with a known response could be sent after the extension negotiation message, eliminating the need for a timeout, but the best I could think of was to request the place in queue (code 51) for a bogus, randomized filename.

As I'm typing this out I realize it would be way more straightforward to have clients send the negotiation message to each other upon connection, then there's no need for a request/response loop.  This message could be sent immediately after PeerInit/PierceFirewall.

Thoughts?

**Comment by mathiascode** [comment]
Generally, I'm against protocol extensions for a proprietary network like Soulseek, and i've deprecated the Nicotine+-specific peer message that was introduced a long time ago: https://github.com/nicotine-plus/nicotine-plus/blob/3bf5613c6b8d32840bfb7936cd44049426e8a5b8/pynicotine/slskmessages.py#L2479-L2483

I don't think it's our position to make these kind of changes, as third party Soulseek clients have already been a controversy ~15 years ago due to malicious and erroneous behavior.

Since Nir specifically mentioned changes to the file transfer protocol as a concern in the past (https://www.slsknet.org/news/comment/718#comment-718), and it's one of the most fragile parts of the protocol, I don't think this is something I want to implement in Nicotine+, unless it's added to the official client.

**Reply by mathiascode** [reply]
Something else that came to mind: concurrent transfers could possibly reduce the need for Soulseek privileges (faster internet speeds have already done this to an extent). These donations are (I would assume) the only thing still keeping the Soulseek server alive today.

In any case, my previous opinion still stands. This change is outside my comfort zone.
